README
libstomp
--------

libstomp - a free implementation of the stomp protocol than can be plugged
to different connection implementations using an adapter interface.

stomp: https://stomp.github.io/

1 adapters available:

 * websockets: make connection to Websockets via https://libwebsockets.org

Based on the Javascript implementation of http://www.jmesnil.net/stomp-websocket/doc/

See exampleProgram as an example of how to connect to a ws endpoint

[Changelogs](ChangeLog)

NEWS
Sample NEWS file for stomp project.


include/libstomp.h
/*
/*
 * libstomp - a free implementation of the stomp protocol than can be plugged
 * to different connection implementations using an adapter interface.
 *
 * 1 adapters available:
 *
 *  * websockets: make connection to Websockets via libwebsockets.
 *
 * https://stomp.github.io/
 * https://github.com/warmcat/libwebsockets
 *
 * Copyright (C) 2017 Sergio Otero <sergio.otero@gmail.com>
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation:
 *  version 2.1 of the License.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 *  MA  02110-1301  USA
 */

#ifndef libstomp_H
#define libstomp_H

#include <time.h>

extern const int STOMP_DEBUG;

#define stomp_debug_print(fmt, ...) \
		do { if (STOMP_DEBUG) fprintf(stderr, fmt, ##__VA_ARGS__); } while (0)

#define STOMP_MAX_FRAME_BUFFER 2048

enum StompAdapterStatus {
	created,
	initialized,
	preconnected,
	connected,
	disconnected,
	destroyed
};

typedef struct {
  char *name;
  char *value;
} StompHeader;

typedef struct {
  size_t len;
  StompHeader *header_array;
} StompHeaders;

typedef struct {
  char *command;
  StompHeaders *system_headers;
  StompHeaders *user_headers;
  char *body;
} StompFrame;

typedef struct StompInfo StompInfo;
typedef struct StompSubscription StompSubscription;


typedef void (*stomp_callback)(StompInfo *stomp_info, const StompFrame *frame);

struct StompSubscription{
  char subscription_id[30]; //TODO length?
  stomp_callback message_callback;
  StompSubscription *previous;
  StompSubscription *next;
};

typedef struct StompAdapter StompAdapter;

typedef int (*stomp_adapter_init_function)(StompAdapter *adapter, StompAdapter *parent_adapter);
typedef int (*stomp_adapter_service_function)(StompAdapter *adapter, int timeout_ms);
typedef int (*stomp_adapter_connect_function)(StompAdapter *adapter);
typedef int (*stomp_adapter_send_function)(StompAdapter *adapter, char *message);
typedef int (*stomp_adapter_restart_function)(StompAdapter *adapter);
typedef int (*stomp_adapter_destroy_function)(StompAdapter *adapter);

typedef int (*stomp_adapter_onopen_callback)(StompAdapter *adapter);
typedef int (*stomp_adapter_onmessage_callback)(StompAdapter *adapter, char *message);
typedef int (*stomp_adapter_onerror_callback)(StompAdapter *adapter, char *message);
typedef int (*stomp_adapter_onheartbeat_callback)(StompAdapter *adapter);
typedef int (*stomp_adapter_onclose_callback)(StompAdapter *adapter, char *message);

struct StompAdapter{
	enum StompAdapterStatus status;
	stomp_adapter_init_function init_function;
	stomp_adapter_connect_function connect_function;
	stomp_adapter_send_function send_function;
	stomp_adapter_service_function service_function;
	stomp_adapter_restart_function restart_function;
	stomp_adapter_destroy_function destroy_function;

	stomp_adapter_onopen_callback onopen_callback;
	stomp_adapter_onmessage_callback onmessage_callback;
	stomp_adapter_onerror_callback onerror_callback;
	stomp_adapter_onheartbeat_callback onheartbeat_callback;
	stomp_adapter_onclose_callback onclose_callback;

	StompAdapter *parent_adapter;
	StompAdapter *child_adapter;

	void *custom_data;
};


struct StompInfo {
	StompAdapter adapter;
	StompHeaders connect_headers;

	stomp_callback connect_callback;
	stomp_callback error_callback;

	int next_subscription_id;
	StompSubscription *subscriptions;

	time_t last_server_action_time;
	void *custom_data;
};

extern StompAdapter stomp_libwebsockets_adapter(char *url);

extern StompInfo stomp_create(StompAdapter *adapter);

extern int stomp_init(StompInfo *stomp_info);

extern int stomp_connect(StompInfo *stomp_info, StompHeaders *headers, stomp_callback connect_callback, stomp_callback error_callback);

extern int stomp_reconnect(StompInfo *stomp_info);

extern StompHeader* stomp_find_header(StompHeaders *headers, char *name);

extern char* stomp_subscribe(StompInfo *stomp_info, char *destination, stomp_callback message_callback, StompHeaders* headers);

extern int stomp_unsubscribe(StompInfo *stomp_info, char *subscription_id);

extern int stomp_service(StompInfo *stomp_info, int timeout_ms);

extern int stomp_destroy(StompInfo *stomp_info);

extern StompAdapter stomp_libwebsockets_adapter(char *url);

#endif


libstomp/libstomp.c
/*
 * libstomp - a free implementation of the stomp protocol than can be plugged
 * to different connection implementations using an adapter interface.
 *
 * 1 adapters available:
 *
 *  * websockets: make connection to Websockets via libwebsockets.
 *
 * https://stomp.github.io/
 * https://github.com/warmcat/libwebsockets
 *
 * Copyright (C) 2017 Sergio Otero <sergio.otero@gmail.com>
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation:
 *  version 2.1 of the License.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 *  MA  02110-1301  USA
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
//
//#include <syslog.h>
//#include <time.h>
//#include <unistd.h>

#include "libstomp.h"

const int STOMP_DEBUG = 1;

typedef struct {
	StompInfo *stomp_info;
} StompAdapterStompInfo;

static StompAdapterStompInfo* get_adapter_custom_data(StompAdapter *adapter) {
	return (StompAdapterStompInfo*)adapter->custom_data;
}

StompHeader* stomp_find_header(StompHeaders *headers, char *name) {
	if (headers == NULL) return NULL;

	for (int i = 0; i < headers->len; i++) {
		if (!strcmp(headers->header_array[i].name, name)) return &headers->header_array[i];
	}

	return NULL;
}

StompSubscription* stomp_find_subscription(StompInfo *stomp_info, char *subscription_id) {
	StompSubscription *current = stomp_info->subscriptions;
	while (current != NULL) {
		if (!strcmp(current->subscription_id, subscription_id)) return current;
		current = current->next;
	}

	return NULL;
}

int stomp_frame_header_marshall(StompHeaders *headers, char *buffer, int skipContentLength) {
	if (headers) {
		for (int i = 0; i < headers->len; i++) {
			StompHeader *header = &headers->header_array[i];

			if (!strcmp(header->name, "content-length")) {
				skipContentLength = !strcmp(header->value, "false");
				continue;
			}
			strcat(buffer, header->name);
			strcat(buffer, ":");
			strcat(buffer, header->value);
			strcat(buffer, "\n");
		}
	}

	return skipContentLength;
}

void stomp_frame_marshall(StompFrame *frame, char *buffer, int maxLength) {
	//TODO maxlength

	strcpy(buffer, frame->command);
	strcat(buffer, "\n");

	int skipContentLength = stomp_frame_header_marshall(frame->system_headers, buffer, skipContentLength);
		skipContentLength = stomp_frame_header_marshall(frame->user_headers, buffer, skipContentLength);

	if (frame->body && !skipContentLength) {
		char temp[25];
		//FIXME it should be utf-8 length, not byte length
		sprintf(temp,"content-length:%lu\n", strlen(frame->body));
		strcat(buffer, temp);
	}

	strcat(buffer, "\n");

	if (frame->body) {
		strcat(buffer, frame->body);
	}

//	strcat(buffer, "\0");
}

int stomp_transmit(StompInfo *stomp_info, StompFrame *frame) {
	StompAdapter *child_adapter = stomp_info->adapter.child_adapter;

	//TODO reuse buffer
    char message[STOMP_MAX_FRAME_BUFFER];

    stomp_frame_marshall(frame, message, STOMP_MAX_FRAME_BUFFER);

	stomp_debug_print("stomp sending:\n%s\n", message);

	return child_adapter->send_function(child_adapter, message);
}

int stomp_send_connect(StompInfo *stomp_info) {
	//TODO headers
	StompHeader system_headers_array[2];
	system_headers_array[0].name = "accept-version";
	system_headers_array[0].value = "1.1,1.0";
	system_headers_array[1].name = "heart-beat";
	system_headers_array[1].value = "10000,10000";

	StompHeaders system_headers = {.len = 2 , .header_array = system_headers_array};
	StompFrame frame = {.command = "CONNECT", .system_headers = &system_headers, .user_headers = &stomp_info->connect_headers, .body = NULL};

	return stomp_transmit(stomp_info, &frame);
}

char* stomp_read_line(char *cur_line) {
  char *next_line = strstr(cur_line, "\n");
  if (next_line) {
	  *next_line = '\0';  // temporarily terminate the current line
	  return &next_line[1];
  } else {
	  return NULL;
  }
}

int stomp_frame_unmarshall(char *message, StompFrame *frame) {
	// read command
	char *cur_line = message;
	char *next_line = stomp_read_line(cur_line);
	frame->command = cur_line;

	StompHeaders *headers = malloc(sizeof(StompHeaders));
	headers->header_array = malloc(10*sizeof(StompHeader)); //TODO dynamic
	headers->len = 0;

	// read headers until newline
	while (next_line != NULL) {
		cur_line = next_line;
		next_line = stomp_read_line(cur_line);

		if (strlen(cur_line) == 0) break;

		char *sep = strchr(cur_line, ':');
		if (sep == NULL) return -1;

		// Separate string in 2 with NULL char
		*sep = 0;

		StompHeader *header = &headers->header_array[headers->len];
		header->name = cur_line;
		header->value = &sep[1];

		headers->len++;
	}

	frame->system_headers = headers;

	// read body
	frame->body = next_line;

	return 0;
}

int stomp_unmarshall_frame_free(StompFrame *frame) {
	free(frame->system_headers->header_array);
	free(frame->system_headers);

	return 0;
}

StompHeaders *stomp_prepare_headers(StompHeaders* system_headers, int system_headers_len, StompHeaders* user_headers) {
	int user_headers_len = user_headers == NULL ? 0 : user_headers->len;
	int total_len = system_headers_len + user_headers_len;

	if (total_len != system_headers->len) {
		return NULL;
	}

	memcpy(&system_headers[system_headers_len], user_headers->header_array, user_headers_len * sizeof(StompHeader));

	return system_headers;
}

void stomp_prepare_connect_headers(StompInfo *stomp_info, StompHeaders* connect_user_headers) {
  // store a copy of the user headers
  StompHeaders *connect_headers = &stomp_info->connect_headers;

  connect_headers->len = connect_user_headers->len;

  if (connect_headers->len > 0 && connect_headers != connect_user_headers) {
	  connect_headers->header_array = malloc(sizeof(StompHeader) * connect_headers->len);
	  memcpy(connect_headers->header_array, connect_user_headers->header_array, connect_headers->len * sizeof(StompHeader));
  }
}

int stomp_connect(StompInfo *stomp_info, StompHeaders *headers, stomp_callback connect_callback, stomp_callback error_callback) {
	if (stomp_info->adapter.status != initialized) return -1;

	StompAdapter *child_adapter = stomp_info->adapter.child_adapter;

	stomp_info->connect_callback = connect_callback;
	stomp_info->error_callback = error_callback;

	stomp_prepare_connect_headers(stomp_info, headers);

	stomp_info->adapter.status = preconnected;

	return child_adapter->connect_function(child_adapter);
}

int stomp_unsubscribe(StompInfo *stomp_info, char *subscription_id) {
	if (stomp_info->adapter.status != connected) return -1;

	StompSubscription *subscription = stomp_find_subscription(stomp_info, subscription_id);

	if (subscription == NULL) return -1;

	if (subscription->previous == NULL) {
		stomp_info->subscriptions = subscription->next;
	}
	if (subscription->next != NULL) {
		subscription->next->previous = subscription->previous;
	}

	free(subscription);

	StompHeader system_headers_array[1];
	system_headers_array[0].name = "id";
	system_headers_array[0].value = subscription_id;

	StompHeaders system_headers = {.len = 1 , .header_array = system_headers_array};
	StompFrame frame = {.command = "UNSUBSCRIBE", .system_headers = &system_headers, .user_headers = NULL, .body = NULL};

	return stomp_transmit(stomp_info, &frame);
}

char* stomp_subscribe(StompInfo *stomp_info, char *destination, stomp_callback message_callback, StompHeaders* headers) {
	if (stomp_info->adapter.status != connected) return NULL;

	StompSubscription *subscription = malloc(sizeof(StompSubscription));
	subscription->message_callback = message_callback;

	StompHeader *header_id = stomp_find_header(headers, "id");
	int num_headers = header_id ? 1 : 2;

	StompHeader system_headers_array[num_headers];
	system_headers_array[0].name = "destination";
	system_headers_array[0].value = destination;
	if (header_id) {
		strcpy(subscription->subscription_id, header_id->value);
	} else {
		sprintf(subscription->subscription_id, "sub-%d", stomp_info->next_subscription_id++);

		system_headers_array[1].name = "id";
		system_headers_array[1].value = subscription->subscription_id;
	}

	StompHeaders system_headers = {.len = num_headers , .header_array = system_headers_array};
	StompFrame frame = {.command = "SUBSCRIBE", .system_headers = &system_headers, .user_headers = headers, .body = NULL};

	if (stomp_transmit(stomp_info, &frame)) {
		free(subscription);
		return NULL;
	}

	subscription->next = stomp_info->subscriptions;
	subscription->previous = NULL;
	stomp_info->subscriptions = subscription;
	if (subscription->next != NULL) subscription->next->previous = subscription;

	return subscription->subscription_id;
}

int stomp_service(StompInfo *stomp_info, int timeout_ms) {
	if (stomp_info->adapter.status != preconnected && stomp_info->adapter.status != connected) return -1;

	StompAdapter *child_adapter = stomp_info->adapter.child_adapter;

	return child_adapter->service_function(child_adapter, timeout_ms);
}

int stomp_destroy_internal(StompInfo *stomp_info, int reconnect) {
	if (stomp_info->adapter.status == destroyed) return -1;

	StompAdapter *adapter = &stomp_info->adapter;
	StompAdapter *child_adapter = adapter->child_adapter;

	StompSubscription *subscription = stomp_info->subscriptions;
	while (subscription != NULL) {
		StompSubscription *next_subscription = subscription->next;
		free(subscription);
		subscription = next_subscription;
	}

	stomp_info->subscriptions = NULL;

	if (reconnect) {
		child_adapter->restart_function(child_adapter);

		stomp_info->adapter.status = initialized;
	} else {
		child_adapter->destroy_function(child_adapter);
		free(adapter->custom_data);

		if (stomp_info->connect_headers.len > 0) {
			free(stomp_info->connect_headers.header_array);
		}

		stomp_info->adapter.status = destroyed;
	}


	return 0;
}

int stomp_destroy(StompInfo *stomp_info) {
	return stomp_destroy_internal(stomp_info, 0);
}

int stomp_reconnect(StompInfo *stomp_info) {
	if (stomp_info->adapter.status == destroyed) return -1;

	if (stomp_destroy_internal(stomp_info, 1) != 0) return -1;

	return stomp_connect(stomp_info, &stomp_info->connect_headers, stomp_info->connect_callback, stomp_info->error_callback);
}

static int onopen_callback(StompAdapter *adapter) {
	StompAdapterStompInfo *custom_info = get_adapter_custom_data(adapter);
	StompInfo *stomp_info = custom_info->stomp_info;

	return stomp_send_connect(stomp_info);
}

static int onerror_callback_internal(StompAdapter *adapter, StompFrame *frame) {
	StompAdapterStompInfo *custom_info = get_adapter_custom_data(adapter);
	StompInfo *stomp_info = custom_info->stomp_info;

	adapter->status = disconnected;

	stomp_info->error_callback(stomp_info, frame);

	return 0;
}

static int onerror_callback(StompAdapter *adapter, char *message) {
	StompHeader header_array[1];
	header_array[0].name = "message";
	header_array[0].value = message;

	StompHeaders headers;
	headers.len = 1;
	headers.header_array = header_array;

	StompFrame frame;
	frame.command = "ERROR";
	frame.system_headers = &headers;

	return onerror_callback_internal(adapter, &frame);
}

static int onclose_callback(StompAdapter *adapter, char *message) {
	return onerror_callback(adapter, message);
}


static int onmessage_callback(StompAdapter *adapter, char *message) {
	StompAdapterStompInfo *custom_info = get_adapter_custom_data(adapter);
	StompInfo *stomp_info = custom_info->stomp_info;

	stomp_debug_print("stomp receive '%s'\n", message);

	StompFrame frame;

	if (stomp_frame_unmarshall(message, &frame)) return -1;

	char *command = frame.command;
	int ret;

	if (!strcmp(command, "CONNECTED")) {
		stomp_info->adapter.status = connected;

		stomp_info->connect_callback(stomp_info, &frame);

		ret = 0;
	} else if (!strcmp(command, "MESSAGE")) {
		StompHeader *header_subscription = stomp_find_header(frame.system_headers, "subscription");

		StompSubscription *subscription = stomp_find_subscription(stomp_info, header_subscription->value);
		if (subscription != NULL) {
			subscription->message_callback(stomp_info, &frame);
			ret = 0;
		} else {
			ret = -1;
		}
	} else if (!strcmp(command, "RECEIPT")) {
		ret = -1;
	} else if (!strcmp(command, "ERROR")) {
		onerror_callback_internal(adapter, &frame);
		ret = 0;
	} else {
		fprintf(stderr, "Invalid command %s\n", command);
		onerror_callback(adapter, "invalid stomp command");
		ret = -1;
	}

	stomp_unmarshall_frame_free(&frame);

	return ret;
}


static int onheartbeat_callback(StompAdapter *adapter) {
	return 0;
}


StompInfo stomp_create(StompAdapter *child_adapter) {
	StompInfo stomp_info;

	stomp_info.adapter.status = created;
	stomp_info.adapter.child_adapter = child_adapter;

	stomp_info.adapter.onopen_callback = onopen_callback;
	stomp_info.adapter.onmessage_callback = onmessage_callback;
	stomp_info.adapter.onerror_callback = onerror_callback;
	stomp_info.adapter.onheartbeat_callback = onheartbeat_callback;
	stomp_info.adapter.onclose_callback = onclose_callback;

	StompAdapterStompInfo *custom_data = malloc(sizeof(StompAdapterStompInfo));
	stomp_info.adapter.custom_data = custom_data;

	stomp_info.subscriptions = NULL;
	stomp_info.last_server_action_time = time(NULL);
	stomp_info.next_subscription_id = 0;

	return stomp_info;
}

int stomp_init(StompInfo *stomp_info) {
	StompAdapterStompInfo *custom_data = get_adapter_custom_data(&stomp_info->adapter);
	custom_data->stomp_info = stomp_info;

	stomp_info->adapter.status = initialized;

	StompAdapter *child_adapter = stomp_info->adapter.child_adapter;
	return child_adapter->init_function(child_adapter, &stomp_info->adapter);
}


exampleProgram/exampleProgram.c
/*
 * libstomp - a free implementation of the stomp protocol than can be plugged
 * to different connection implementations using an adapter interface.
 *
 * 1 adapters available:
 *
 *  * websockets: make connection to Websockets via libwebsockets.
 *
 * https://stomp.github.io/
 * https://github.com/warmcat/libwebsockets
 *
 * Copyright (C) 2017 Sergio Otero <sergio.otero@gmail.com>
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation:
 *  version 2.1 of the License.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 *  MA  02110-1301  USA
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <unistd.h>

#include "libstomp.h"

volatile int test_stomp_force_exit = 0;
volatile int test_stomp_force_reconnect = 0;

void test_stomp_sighandler(int sig)
{
	test_stomp_force_exit = 1;
}

void test_stomp_message_callback(StompInfo *stomp_info, const StompFrame *frame) {
	fprintf(stdout, "command %s message %s\n", frame->command, frame->body);
}

void test_stomp_subscribe(StompInfo *stomp_info) {
	char *topic = (char *)stomp_info->custom_data;

	fprintf(stdout, "subscribing %s !\n", topic);

	char *suscription_id = stomp_subscribe(stomp_info, topic, test_stomp_message_callback, NULL);
}

void test_stomp_connect_callback(StompInfo *stomp_info, const StompFrame *frame) {
	fprintf(stdout, "connected %s !\n", frame->command);

	test_stomp_subscribe(stomp_info);
}

void test_stomp_error_callback(StompInfo *stomp_info, const StompFrame *frame) {
	fprintf(stderr, "error callback %s !\n", frame->command);

	StompHeader *header = stomp_find_header(frame->system_headers, "message");

	if (header != NULL && header->value != NULL) {
		fprintf(stderr, "error message %s\n", header->value);
	}
	if (header != NULL && header->value != NULL && strstr(header->value, "AccessDeniedException") != NULL) {
		test_stomp_force_exit = 1;
	} else {
		test_stomp_force_reconnect = 1;
	}

	return;
}


int main(int argc, char **argv) {

  if (argc < 3) goto usage;

  char *fullURL = argv[1];
  char *token = argv[2];
  char *topic = argv[3];

  signal(SIGINT, test_stomp_sighandler);

  fprintf(stdout, "connecting %s !\n", fullURL);

  StompAdapter ws_adapter = stomp_libwebsockets_adapter(fullURL);
  StompInfo stomp_info = stomp_create(&ws_adapter);

  stomp_init(&stomp_info);

  StompHeader header_array[1];
  header_array[0].name = "Authorization";
  header_array[0].value = token;

  StompHeaders headers;
  headers.len = 1;
  headers.header_array = header_array;

  if (stomp_connect(&stomp_info, &headers, test_stomp_connect_callback, test_stomp_error_callback)) {
	fprintf(stderr, "error initializing");
	return 1;
  }

  stomp_info.custom_data = topic;

  while (!test_stomp_force_exit) {
  	stomp_service(&stomp_info, 500);

  	if (test_stomp_force_reconnect) {
  		sleep(1);
  		stomp_reconnect(&stomp_info);
  		test_stomp_force_reconnect = 0;
  	}
  }

  stomp_destroy(&stomp_info);

  return 0;

  usage:
  	fprintf(stderr, "Usage: exampleProgram <wsAddress> <AuthorizationHeader> <topic> \n");
  	return 1;
}


libstomp/stomp_adapter_libwebsockets.c
/*
 * libstomp - a free implementation of the stomp protocol than can be plugged
 * to different connection implementations using an adapter interface.
 *
 * 1 adapters available:
 *
 *  * websockets: make connection to Websockets via libwebsockets.
 *
 * https://stomp.github.io/
 * https://github.com/warmcat/libwebsockets
 *
 * Copyright (C) 2017 Sergio Otero <sergio.otero@gmail.com>
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation:
 *  version 2.1 of the License.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 *  MA  02110-1301  USA
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <libwebsockets.h>

#include "libstomp.h"

typedef struct {
	struct lws *wsi; // Websocket instance
	struct lws_context *context;
	char *url;
} StompAdapterLibWebSocketsData;

static StompAdapterLibWebSocketsData* get_adapter_custom_data(StompAdapter *adapter) {
	return (StompAdapterLibWebSocketsData*)adapter->custom_data;
}

#define STOMP_DEFLATE_BUFFER "rx_buf_size=11" // 2^n=2048

int stomp_libwebsockets_callback_lws_http(struct lws *wsi, enum lws_callback_reasons reason,
			void *user, void *in, size_t len);

int stomp_libwebsockets_callback_lws_websocket(struct lws *wsi, enum lws_callback_reasons reason,
			void *user, void *in, size_t len);

/* list of supported protocols and callbacks */
enum stomp_lws_protocol_enum {
		PROTOCOL_HTTP,
		PROTOCOL_STOMP12
	};

static const struct lws_protocols stomp_lws_protocols[] = {
    // first protocol must always be HTTP handler
    {
        "http-only",                // name
		stomp_libwebsockets_callback_lws_http,          // callback
        0,                          // per_session_data_size
		STOMP_MAX_FRAME_BUFFER // rx_buffer_size
    },
    {
        "v12.stomp",                // protocol name - very important!
		stomp_libwebsockets_callback_lws_websocket,      // callback
        0,                           // we don't use any per session data
		STOMP_MAX_FRAME_BUFFER // rx_buffer_size
    },
    {
        NULL, NULL, 0               // End of list
    }
};

static const struct lws_extension stomp_lws_exts[] = {
	{
		"permessage-deflate",
		lws_extension_callback_pm_deflate,
		"permessage-deflate; client_no_context_takeover"
	},
	{
		"deflate-frame",
		lws_extension_callback_pm_deflate,
		"deflate_frame"
	},
	{ NULL, NULL, NULL /* terminator */ }
};

static int init_function(StompAdapter *adapter, StompAdapter *parent_adapter) {
	if (adapter->status != created) return -1;

	adapter->parent_adapter = parent_adapter;

	adapter->status = initialized;

	return 0;
}

static int connect_function (StompAdapter *adapter) {
	if (adapter->status != initialized) return -1;

	int use_ssl = 0, ret = 0, ietf_version = -1;
	unsigned int pp_secs = 0;
	struct lws_context_creation_info info;
	struct lws_client_connect_info i;

	const char *prot, *p, *url;
	char fullURL[300];
	char path[300];
	char token[300];
	char deviceId[300];

	StompAdapterLibWebSocketsData *custom_data = get_adapter_custom_data(adapter);
	url = custom_data->url;

	memset(&info, 0, sizeof info);

	memset(&i, 0, sizeof(i));

	// Copiamos url a fullURL pq la funcion toca el string
	strncpy(fullURL, url, sizeof(fullURL) - 1);
	if (lws_parse_uri(fullURL, &prot, &i.address, &i.port, &p)) {
		fprintf(stderr, "Error parsing URL %s\n", url);
		return -1;
	}

	/* add back the leading / on path */
	path[0] = '/';
	strncpy(path + 1, p, sizeof(path) - 2);
	path[sizeof(path) - 1] = '\0';
	i.path = path;

	if (!strcmp(prot, "http") || !strcmp(prot, "ws")) {
		use_ssl = 0;
	}
	if (!strcmp(prot, "https") || !strcmp(prot, "wss")) {
		if (!use_ssl) {
			use_ssl = LCCSCF_USE_SSL;
		}
	}

	/*
	* create the websockets context.  This tracks open connections and
	* knows how to route any traffic and which protocol version to use,
	* and if each connection is client or server side.
	*
	* We tell it to not listen on any port.
	*/

	info.port = CONTEXT_PORT_NO_LISTEN;
	info.protocols = stomp_lws_protocols;
	info.gid = -1;
	info.uid = -1;
	info.ws_ping_pong_interval = pp_secs;
	info.extensions = stomp_lws_exts;
	info.max_http_header_data = 2048;

#if defined(LWS_OPENSSL_SUPPORT)
	info.options |= LWS_SERVER_OPTION_DO_SSL_GLOBAL_INIT;
#endif

	custom_data->context = lws_create_context(&info);
	if (custom_data->context == NULL) {
		fprintf(stderr, "Creating libwebsocket context failed\n");
		return -1;
	}

	i.context = custom_data->context;
	i.ssl_connection = use_ssl;
	i.host = i.address;
	i.origin = i.address;
	i.ietf_version_or_minus_one = ietf_version;

	stomp_debug_print("using %s mode (ws)\n", prot);

	/*
	 * nothing happens until the client websocket connection is
	 * asynchronously established... calling lws_client_connect() only
	 * instantiates the connection logically, lws_service() progresses it
	 * asynchronously.
	 */

	stomp_debug_print("Opening socket \n");
	i.protocol = stomp_lws_protocols[PROTOCOL_STOMP12].name;
	i.pwsi = &custom_data->wsi;

	i.userdata = adapter;

	struct lws *result = lws_client_connect_via_info(&i);
	if (!result) {
		fprintf(stderr, "Error opening socket!\n");
		return -1;
	}

	adapter->status = preconnected;

	return 0;
}

static int send_function (StompAdapter *adapter, char *message) {
	if (adapter->status != connected) return -1;

	StompAdapterLibWebSocketsData *custom_data = get_adapter_custom_data(adapter);

	//TODO reuse buffer
	char buffer[LWS_PRE + STOMP_MAX_FRAME_BUFFER];

	int message_len = strlen(message) + 1; // send the null char

	if (strlen(message) > STOMP_MAX_FRAME_BUFFER) {
		fprintf(stderr, "message exceed STOMP_MAX_FRAME_BUFFER %d > %d", message_len, STOMP_MAX_FRAME_BUFFER);
		return -1;
	}

	strcpy(&buffer[LWS_PRE], message);

	int n = lws_write(custom_data->wsi, (unsigned char *)&buffer[LWS_PRE], message_len, LWS_WRITE_TEXT);
	if (n < 0)
		return -1;
	/* we only had one thing to send, so inform lws we are done
	 * if we had more to send, call lws_callback_on_writable(wsi);
	 * and just return 0 from callback.  On having sent the last
	 * part, call the below api instead.*/
	lws_client_http_body_pending(custom_data->wsi, 0);

	return 0;

}

static int service_function (StompAdapter *adapter, int timeout_ms) {
	if (adapter->status != preconnected && adapter->status != connected) return -1;

	StompAdapterLibWebSocketsData *custom_data = get_adapter_custom_data(adapter);

	int status = lws_service(custom_data->context, timeout_ms);
	if (status != 0) {
		fprintf(stderr, "Status is %d!!!!!!\n", status);
	}
	return status;
}

static int destroy_function_internal (StompAdapter *adapter, int reconnect) {
	if (adapter->status == destroyed) return -1;

	StompAdapterLibWebSocketsData *custom_data = get_adapter_custom_data(adapter);

	if (custom_data->context) {
		lws_context_destroy(custom_data->context);
	}

	if (reconnect) {
		adapter->status = initialized;
	} else {
		free(adapter->custom_data);
		adapter->status = destroyed;
	}

	return 0;
}

static int destroy_function (StompAdapter *adapter) {
	return destroy_function_internal(adapter, 0);
}

static int restart_function(StompAdapter *adapter) {
	if (adapter->status == destroyed) return -1;

	if (destroy_function_internal(adapter, 1) != 0) return -1;

	adapter->status = initialized;

	return 0;
}

StompAdapter stomp_libwebsockets_adapter(char *url) {
	StompAdapter adapter;

	adapter.status = created;
	adapter.init_function = init_function;
	adapter.service_function = service_function;
	adapter.connect_function = connect_function;
	adapter.send_function = send_function;
	adapter.restart_function = restart_function;
	adapter.destroy_function = destroy_function;

	StompAdapterLibWebSocketsData *custom_data = malloc(sizeof(StompAdapterLibWebSocketsData));
	custom_data->url = url;

	adapter.custom_data = custom_data;

	return adapter;
}

int stomp_libwebsockets_callback_lws_http(struct lws *wsi, enum lws_callback_reasons reason,
			void *user, void *in, size_t len) {
	StompAdapter *adapter = (StompAdapter *)user;
	StompAdapter *parent_adapter = adapter != NULL ? adapter->parent_adapter : NULL;
	char *message = (char *)in;

	switch (reason) {
		case LWS_CALLBACK_CLIENT_CONNECTION_ERROR:
			if (adapter->status != preconnected && adapter->status != connected) return 0;

			parent_adapter->onclose_callback(parent_adapter, message);
			break;
	}

	return 0;
}


int stomp_libwebsockets_callback_lws_websocket(struct lws *wsi, enum lws_callback_reasons reason,
			void *user, void *in, size_t len)
{
	StompAdapter *adapter = (StompAdapter *)user;
	StompAdapter *parent_adapter = NULL;

	if (adapter != NULL) {
		parent_adapter = adapter->parent_adapter;
	}

	char *message = (char *)in;

	stomp_debug_print("stomp_callback http %i !!\n", reason);
	switch (reason) {
		case LWS_CALLBACK_WS_EXT_DEFAULTS:
			// En este momento se cambia el tamaño del buffer de compresion para que no se partan los mensajes en 2 callbacks
			if (!strcmp(user, "permessage-deflate")) {
				strcpy(in, STOMP_DEFLATE_BUFFER);
			}
			break;
		case LWS_CALLBACK_CLIENT_ESTABLISHED:
			if (adapter->status != connected && adapter->status != preconnected) return 0;

			adapter->status = connected;

			parent_adapter->onopen_callback(parent_adapter);
			break;
		case LWS_CALLBACK_CLIENT_RECEIVE:
			if (adapter->status != connected && adapter->status != preconnected) return 0;

			if (len == 0) return 0;

			message[len] = '\0';

			parent_adapter->onmessage_callback(parent_adapter, message);
			break;
		case LWS_CALLBACK_CLOSED:
			if (adapter->status != connected && adapter->status != preconnected) return 0;

			adapter->status = disconnected;

			parent_adapter->onclose_callback(parent_adapter, message);
			break;
		default:
			break;
	}

	return 0;
}
